pilot/pkg/xds/workloadentry.go

```diff
@@ -0,0 +1,252 @@
+	if !features.WorkloadEntryAutoRegistration {
+		return
+	}
+	ticker := time.NewTicker(10 * features.WorkloadEntryCleanupGracePeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			wles, err := sg.Store.List(gvk.WorkloadEntry, metav1.NamespaceAll)
+			if err != nil {
+				adsLog.Warnf("error listing WorkloadEntry for cleanup: %v", err)
+				continue
+			}
+			for _, wle := range wles {
+				wle := wle
+				if !shouldCleanupEntry(wle) {
+					continue
+				}
+				sg.cleanupQueue.Push(func() error {
+					sg.cleanupEntry(wle)
+					return nil
+				})
+			}
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (sg *InternalGen) cleanupEntry(wle config.Config) {
+	if err := sg.cleanupLimit.Wait(context.TODO()); err != nil {
+		adsLog.Errorf("error in WorkloadEntry cleanup rate limiter: %v", err)
+	}
+	if err := sg.Store.Delete(gvk.WorkloadEntry, wle.Name, wle.Namespace); err != nil {
+		adsLog.Warnf("failed cleaning up auto-registered WorkloadEntry %s/%s: %v", wle.Namespace, wle.Name, err)
+	}
+}
+
+func shouldCleanupEntry(wle config.Config) bool {
+	// don't clean-up if connected or non-autoregistered WorkloadEntries
+	_, ok := wle.Annotations[WorkloadControllerAnnotation]
+	if wle.Annotations[AutoRegistrationGroupAnnotation] == "" || ok {
+		return false
+	}
+
+	disconnUnixTime, err := strconv.Atoi(wle.Annotations[DisconnectedAtAnnotation])
+	if err != nil {
+		// remove workload entries with invalid disconnect times - they need to be re-registered and fixed.
+		adsLog.Warnf("invalid disconnect time for WorkloadEntry %s/%s: %s", wle.Annotations[DisconnectedAtAnnotation])
+	}
+	disconnAt := time.Unix(0, int64(disconnUnixTime))
+	// if we haven't passed the grace period, don't cleanup
+	if err == nil && time.Since(disconnAt) < features.WorkloadEntryCleanupGracePeriod {
+		return false
+	}
+
+	return true
+}
+
+func setConnectMeta(c *config.Config, controller string, con *Connection) {
+	c.Annotations[WorkloadControllerAnnotation] = controller
+	c.Annotations[ConnectedAtAnnotation] = strconv.FormatInt(con.Connect.UnixNano(), 10)
+}
+
+func workloadEntryFromGroup(name string, proxy *model.Proxy, groupCfg *config.Config) *config.Config {
+	group := groupCfg.Spec.(*v1alpha3.WorkloadGroup)
+	entry := group.Template.DeepCopy()
+	entry.Address = proxy.IPAddresses[0]
+	// TODO move labels out of entry
+	entry.Labels = mergeLabels(entry.Labels, proxy.Metadata.Labels)
+
+	if proxy.Metadata.Network != "" {
+		entry.Network = proxy.Metadata.Network
+	}
+	if proxy.Locality != nil {
+		entry.Locality = util.LocalityToString(proxy.Locality)
+	}
+	return &config.Config{
+		Meta: config.Meta{
+			GroupVersionKind: gvk.WorkloadEntry,
+			Name:             name,
+			Namespace:        proxy.Metadata.Namespace,
+			Labels:           entry.Labels,
+			Annotations:      map[string]string{AutoRegistrationGroupAnnotation: groupCfg.Name},
+		},
+		Spec: entry,
+		// TODO status fields used for garbage collection
+		Status: nil,
```

Comment:
what would status be populated with?

Reply:
The annotations will eventually be put in a proper API like status. Opted out of doing this to avoid dealing with the Conditions slice. Will probably change this down the line.